Fix radio group validity update when removing or selecting an input

elomscansio opened the pull request:
This PR fixes an issue where radio inputs in the same group failed to correctly update their `validity.valueMissing` state when:

- A **checked radio button was removed** from the DOM.
- A **different radio button was selected** by user interaction.

This behavior caused mismatches with how browsers like Firefox handle radio group validation.

---

### Changes in This PR

#### Radio group revalidation on DOM removal
- Updated `unbind_from_tree()` to revalidate other radio buttons in the same group when a checked input is removed.
- Uses `UnbindContext::parent` as the DOM root to ensure the correct context is used during traversal.

#### New helper: `find_related_radios()`
- Encapsulates logic for finding other inputs in the same group.
- Used during both removal and attribute changes for consistency.

#### Validation on `checked`/`value` updates
- Introduced `update_related_validity_states()` to revalidate all group members when a radio's `checked` or `value` is changed.

#### Web Platform Test (WPT) coverage
- Created a new WPT file: `radio-group-valueMissing.html`.
- Tests follow recommended `test()` pattern:
  - **Precondition**: Assert initial `valueMissing`.
  - **Action**: Remove or select a radio.
  - **Postcondition**: Assert expected `valueMissing`.

#### Manifest updated
- The WPT manifest now includes the new test.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #36110

<!-- Either: -->
- [X] There are tests for these changes

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51731) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51731).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51731).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51731).

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51740) with upstreamable changes.

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#51740) title and body.

elomscansio commented:
hello @jdm there is an error cause by new update https://github.com/servo/servo/actions/runs/14183667086/job/39734917312?pr=36252#step:9:1241 should i use `CanGc::note()` for the `perform_validation_and_update` or it should come from caller function if so how do caller function get its can_gc arg..

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

elomscansio commented:
> hello @jdm there is an error cause by new update https://github.com/servo/servo/actions/runs/14183667086/job/39734917312?pr=36252#step:9:1241 should i use `CanGc::note()` for the `perform_validation_and_update` or it should come from caller function if so how do caller function get its can_gc arg..

oh i have added can_gc to where possible and make sure it compiles

jdm commented on tests/wpt/tests/html/semantics/forms/constraints/form-validation-validity-valueMissing.html:
```diff
@@ -12,6 +12,45 @@
+<input type="radio" id="fourth" name="group1" />
+
+<script>
+  const first = document.getElementById("first");
```
This test file already has a common structure for all of the tests that it runs. Rather than adding new subtests that use a different structure, we should either fit the tests into the existing structure or move them into a new file where we can use whatever rest structure we want.

scansio replied:
creating new test file would be better, because of the remove and click action. If so, what would be the filename.

jdm replied:
radio-group-valueMissing.html should be fine.

jdm commented on components/script/dom/node.rs:
```diff
@@ -381,6 +381,10 @@ impl Node {
 
+        if let Some(input_element) = child.downcast::<HTMLInputElement>() {
+            input_element.pre_remove(can_gc);
+        }
```
I would like to avoid adding element-specific behavior here, since this is an algorithm that's defined by the DOM specification and this change deviates from that. If we implement my original suggestion, all of the logic changes live inside code in htmlinputelement.rs.

jdm replied:
Huh. I'll need some more time to read more code tomorrow.

elomscansio replied:
> Huh. I'll need some more time to read more code tomorrow.

Hello @jdm just a ping for to know if later got time to read the code, I did like to take on another issue

jdm replied:
Please do continue working on other things while you wait for me! I have been delayed by childcare.

elomscansio replied:
alright, take your time sir.

jdm replied:
Ok, I see what you mean! I misread radio_node_iter, but I see a way to address this problem. [UnbindContext](https://doc.servo.org/script/dom/node/struct.UnbindContext.html) includes a `parent` field which is the parent of the node that was removed from the tree. That means that we can change the `elem.upcast::<Node>().GetRootNode(..)` call to use a new argument that we can pass in. In most callers that will be `elem.upcast::<Node>()`, but for the unbind_from_tree case it can be `context.parent`. Then we can use my earlier suggestion of storing the `self.form_owner()` value and passing that in as well.

I recommend leaving `radio_node_iter` as it is and extracting most of its body into a new method that unbind_from_context can call, which will avoid needing to change any other callers of radio_node_iter.

elomscansio replied:
I have made the requested changes

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -1704,6 +1703,25 @@ fn broadcast_radio_checked(broadcaster: &HTMLInputElement, group: Option<&Atom>)
+    }
+}
+
+fn perform_validation_conditionaly(elem: &HTMLInputElement, can_gc: CanGc) {
```
Let's name this `update_related_validity_states`. Right now the name sounds like we may not update anything at all.

jdm commented on tests/wpt/tests/html/semantics/forms/constraints/form-validation-validity-valueMissing.html:
```diff
@@ -12,6 +12,45 @@
+  const third = document.getElementById("third");
+  const fourth = document.getElementById("fourth");
+
+  test(() => {
```
The recommended usage of the test() function is:
```
test(() => {
  // Test preconditions: state of valueMissing before any changes
  // Modify the tree
  // Test post conditions: state of valueMissing after the changes
}, "valueMissing is true for all group members when checked group member is removed");
```

This makes it easier to understand the overall behavior being tested, and ensures that any exceptions are caught and reported appropriately.

jdm reviewed (changes requested):
This is a good start! I appreciate that you added the new test.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#51740) title and body.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#51740) title and body.

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -1704,6 +1703,38 @@ fn broadcast_radio_checked(broadcaster: &HTMLInputElement, group: Option<&Atom>)
+    form: Option<&'a HTMLFormElement>,
+    group: Option<&'a Atom>,
+) -> impl Iterator<Item = DomRoot<HTMLInputElement>> + 'a {
+    let root = context.parent.GetRootNode(&GetRootNodeOptions::empty());
```
Instead of duplicate radio_group_iter, can we make `radio_group_iter` call `find_related_radios`? This requires passing a `&'a Node` argument instead of `context`.

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -2549,13 +2580,13 @@ impl VirtualMethods for HTMLInputElement {
 
     fn unbind_from_tree(&self, context: &UnbindContext, can_gc: CanGc) {
+        let form = self.form_owner();
+        let group = self.radio_group_name();
```
We should not need to store the group name here. We can call it later without any issues.

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -2569,8 +2600,15 @@ impl VirtualMethods for HTMLInputElement {
 
-        self.validity_state()
-            .perform_validation_and_update(ValidationFlags::all(), can_gc);
+        if self.input_type() == InputType::Radio && self.Checked() {
```
I don't think it's enough to check for `self.Checked()`—this radio input could have `self.Required()`, so removing it could still change the valueMissing value for its siblings. Let's only check the `input_type` instead.

jdm commented on tests/wpt/tests/html/semantics/forms/constraints/form-validation-validity-valueMissing.html:
```diff
@@ -12,6 +12,45 @@
+  test(() => {
+    assert_equals(fourth.validity.valueMissing, false);
+  }, "The valueMissing of fourth radio should be false because some in its group were checked");
+</script>
```
These changes need to be removed from this file.

jdm replied:
This comment still needs to be addressed.

scansio replied:
i forget to revert it. I just did that

jdm commented on tests/wpt/tests/html/semantics/forms/constraints/radio-group-valueMissing.html:
```diff
@@ -0,0 +1,46 @@
+  The constraint validation API Test: element.validity.valueMissing for radio
+  group
+</title>
+<link rel="author" title="Intel" href="http://www.intel.com/" />
```
You should either remove this line or change it to add your name and email, like https://github.com/web-platform-tests/wpt/blob/5ff2643aa132c76bb9506a3becb59e2686fd34ec/css/css-lists/counter-slot-order.html#L5

jdm reviewed (changes requested):
Nearly there!

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -1681,19 +1680,15 @@ impl HTMLInputElementMethods<crate::DomTypeHolder> for HTMLInputElement {
+fn radio_group_iter(
+    elem: &HTMLInputElement,
+    group: Option<&Atom>,
+) -> Vec<DomRoot<HTMLInputElement>> {
```
Why the change to the return type?

elomscansio replied:
We initially had `radio_group_iter` return an iterator directly, but that caused borrow checker issues due to the returned iterator capturing references to temporary values like the root node and the form.

To resolve lifetime and ownership issues — particularly around: https://github.com/elomscansio/servo/blob/ebbbcf0dcff8ecf0b1e8e74c5b475242a1a4d063/components/script/dom/htmlinputelement.rs#L1683-L1691

we switched the return type to `Vec<DomRoot<HTMLInputElement>>` and collected the results early. This way, we avoid returning an iterator that borrows from local variables, which the compiler flagged as invalid due to the short-lived nature of those values.

Also, this gives more flexibility at the call sites, https://github.com/elomscansio/servo/blob/ebbbcf0dcff8ecf0b1e8e74c5b475242a1a4d063/components/script/dom/htmlinputelement.rs#L2873-L2875, without running into lifetime or temporary value problems.

elomscansio replied:
I actually did
```rust
fn radio_group_iter<'a>(
    elem: &'a HTMLInputElement,
    group: Option<&'a Atom>,
) -> impl Iterator<Item = DomRoot<HTMLInputElement>> + 'a {
    // Store root in a variable to extend its lifetime
    let root_node = elem
        .upcast::<Node>()
        .GetRootNode(&GetRootNodeOptions::empty());

    let form = elem.form_owner(); // avoid inline temporary
    find_related_radios(&root_node, elem, form.as_deref(), group)
}
```

But got
```bash

(servo) elomscansio@fedora:~/servo$ ./mach build -d
   Compiling script v0.0.1 (/home/elomscansio/servo/components/script)
warning: unknown lint: `crown_is_not_used`
  --> components/script_bindings/lib.rs:11:9
   |
11 | #![deny(crown_is_not_used)]
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unknown_lints)]` on by default

warning: `script_bindings` (lib) generated 1 warning
   Compiling servoshell v0.0.1 (/home/elomscansio/servo/ports/servoshell)
warning: unknown lint: `crown_is_not_used`
  --> components/script/lib.rs:13:9
   |
13 | #![deny(crown_is_not_used)]
   |         ^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unknown_lints)]` on by default

error[E0515]: cannot return value referencing local variable `form`
    --> components/script/dom/htmlinputelement.rs:1692:5
     |
1692 |     find_related_radios(&root_node, elem, form.as_deref(), group)
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----^^^^^^^^^^^^^^^^^^^
     |     |                                     |
     |     |                                     `form` is borrowed here
     |     returns a value referencing data owned by the current function
     |
     = help: use `.collect()` to allocate the iterator

error[E0515]: cannot return value referencing local variable `root_node`
    --> components/script/dom/htmlinputelement.rs:1692:5
     |
1692 |     find_related_radios(&root_node, elem, form.as_deref(), group)
     |     ^^^^^^^^^^^^^^^^^^^^----------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |     |                   |
     |     |                   `root_node` is borrowed here
     |     returns a value referencing data owned by the current function
     |
     = help: use `.collect()` to allocate the iterator

For more information about this error, try `rustc --explain E0515`.
warning: `script` (lib) generated 1 warning
error: could not compile `script` (lib) due to 2 previous errors; 1 warning emitted
      Timing report saved to /home/elomscansio/servo/target/cargo-timings/cargo-timing-20250405T133436Z.html
Failed in 0:07:18
[Warning] Could not generate notification: No module named 'dbus'
```
I just used compiler suggestion

jdm replied:
I believe this error will go away if you use `Option<DomRoot<HTMLFormElement>>` instead of `Option<&'a HTMLFormElement>`.

elomscansio replied:
> I believe this error will go away if you use `Option<DomRoot<HTMLFormElement>>` instead of `Option<&'a HTMLFormElement>`.

I tried it but still got error.
So i thought it would be better to get rid of `find_related_radios` that was introduced and add extra argument to 
```rust
fn radio_group_iter<'a>(
    elem: &'a HTMLInputElement,
    group: Option<&'a Atom>,
    form: Option<&'a HTMLFormElement>,
    root: &'a Node,
) -> impl Iterator<Item = DomRoot<HTMLInputElement>> + 'a {
    root.traverse_preorder(ShadowIncluding::No)
        .filter_map(DomRoot::downcast::<HTMLInputElement>)
        .filter(move |r| &**r == elem || in_same_group(r, form, group, None))
}
``` root and form_group. This avoids borrow checker issue

elomscansio committed:
fix(form-validation): update radio group validity on input removal and selection

This fixes an issue where radio buttons in the same group failed to properly
update their `validity.valueMissing` state in two scenarios:
1. When a checked radio button was removed from the DOM.
2. When a different radio button was selected by the user.

Previously, Servo did not revalidate other radio inputs in the group,
causing discrepancies in `valueMissing` behavior compared to Firefox.

- **Group-wide validation during removal**
  - Updated `unbind_from_tree()` to trigger validation of related radio buttons
    when a checked radio is removed.
  - Uses `UnbindContext::parent` to obtain the correct DOM root context
    during removal.

- **Helper for group traversal**
  - Reuses the `parent` node from `UnbindContext` to iterate over group members
    and trigger `perform_validation_and_update()` as needed.

- **Validation on attribute changes**
  - Added `update_related_validity_states()` to revalidate all related radios
    when `checked` or `value` attributes change.
  - Ensures consistent group validation behavior on user interaction.

- **Web Platform Test coverage**
  - Created a new test file `radio-group-valueMissing.html` to isolate scenarios.
  - Tests follow the recommended structure:
    - **Preconditions:** Validate initial `valueMissing` state.
    - **Actions:** Remove or select radio inputs.
    - **Postconditions:** Assert expected `valueMissing` results.

- **WPT Manifest**
  - Added the new test to the manifest.

Also ensured proper use of `can_gc` where needed to support safe GC operations.

Fixes: [#36110](https://github.com/servo/servo/issues/36110)

Signed-off-by: Emmanuel Elom <elomemmanuel007@gmail.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

jdm reviewed (approved):
Looks good! Thanks for fixing this and adding the new tests.

jdm commented:
```
OK /html/semantics/forms/the-input-element/radio-disconnected-group-owner.html

    FAIL [expected PASS] subtest: Disconnected radio buttons should be contained by their tree root.

    assert_true: Element should suffer from value missing expected true got false
```
Oops! In the unbind case we still need to call `self.validity_state().perform_validation_and_update(...)` because the node isn't part of the group any more. Maybe `update_related_validity_states` should have a special case for radio buttons and then always call the other method as well?

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51740).
